onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc

```diff
@@ -118,36 +129,59 @@ void Model::ResetExecution() {
   prepared_for_exe_ = false;
 }
 
-void Model::Predict() {
-  PrepareForExecution();
+Status Model::Predict(const std::vector<int32_t>& dynamic_outputs, std::vector<Shaper::Shape>& dynamic_output_shapes) {
+  ORT_RETURN_IF_ERROR(PrepareForExecution());
 
   ANeuralNetworksEvent* event = nullptr;
-  THROW_ON_ERROR(nnapi_->ANeuralNetworksExecution_startCompute(execution_, &event));
+  RETURN_STATUS_ON_ERROR(nnapi_->ANeuralNetworksExecution_startCompute(execution_, &event));
 
-  THROW_ON_ERROR(nnapi_->ANeuralNetworksEvent_wait(event));
+  RETURN_STATUS_ON_ERROR(nnapi_->ANeuralNetworksEvent_wait(event));
 
   nnapi_->ANeuralNetworksEvent_free(event);
 
+  dynamic_output_shapes.clear();
+  dynamic_output_shapes.reserve(dynamic_outputs.size());
+  for (const int32_t i : dynamic_outputs) {
+    uint32_t output_rank = 0;
+    RETURN_STATUS_ON_ERROR(nnapi_->ANeuralNetworksExecution_getOutputOperandRank(execution_, i, &output_rank));
+
+    std::vector<uint32_t> output_shape(output_rank);
+    RETURN_STATUS_ON_ERROR(nnapi_->ANeuralNetworksExecution_getOutputOperandDimensions(execution_, i, output_shape.data()));
+
+    dynamic_output_shapes.push_back(output_shape);
+  }
+
   ResetExecution();
+
+  return Status::OK();
 }
 
-void Model::SetInputBuffers(const std::vector<InputBuffer>& inputs) {
-  PrepareForExecution();
+Status Model::SetInputBuffers(const std::vector<InputBuffer>& inputs) {
+  ORT_RETURN_IF_ERROR(PrepareForExecution());
 
   for (size_t i = 0; i < inputs.size(); i++) {
-    SetInputBuffer(i, inputs[i]);
-    shaper_for_exeuction_.UpdateShape(input_names_[i], inputs[i].type.dimensions);
+    ORT_RETURN_IF_ERROR(SetInputBuffer(i, inputs[i]));
+    ORT_RETURN_IF_ERROR(
+        shaper_for_exeuction_.UpdateShape(input_names_[i], inputs[i].type.dimensions));
   }
 
-  shaper_for_exeuction_.UpdateDynamicDimensions();
+  ORT_RETURN_IF_ERROR(shaper_for_exeuction_.UpdateDynamicDimensions());
+
+  return Status::OK();
 }
 
-void Model::SetOutputBuffers(const std::vector<OutputBuffer>& outputs) {
-  PrepareForExecution();
+Status Model::SetOutputBuffers(const std::vector<OutputBuffer>& outputs) {
+  ORT_RETURN_IF_ERROR(PrepareForExecution());
 
```

Comment:
nit: wouldn't have expected calling SetInputBuffers or SetOutputBuffers to have a PrepareForExecution call. feels like you're cluttering the implementation a little vs. telling the user of the class they need to call PrepareForExecution first.

Reply:
Ideally I want to split the model class to 
1. Model class contains NNAPI model and related info
2. Execution class contains only temporary execution related info

Added a TODO for this



---
In reply to: [462109992](https://github.com/microsoft/onnxruntime/pull/4650#discussion_r462109992) [](ancestors = 462109992)

Reply:
Fixed in the new iteration

---
In reply to: [462646759](https://github.com/microsoft/onnxruntime/pull/4650#discussion_r462646759) [](ancestors = 462646759,462109992)